src/config.rs

```diff
@@ -0,0 +1,211 @@
+    ConfiguredMailUser,
+    ConfiguredMailPw,
+    ConfiguredMailPort,
+    ConfiguredSendServer,
+    ConfiguredSendUser,
+    ConfiguredSendPw,
+    ConfiguredSendPort,
+    ConfiguredServerFlags,
+    Configured,
+}
+
+// deprecated
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Display, EnumString, AsRefStr, EnumIter)]
+pub enum SysConfig {
+    #[strum(serialize = "sys.version")]
+    Version,
+    #[strum(serialize = "sys.msgsize_max_recommended")]
+    MsgsizeMaxRecommended,
+    #[strum(serialize = "sys.config_keys")]
+    ConfigKeys,
+}
+
+/// Get a configuration key.
+/// Returns "" when the key is invalid, or no default was found.
+pub fn get(context: &Context, key: impl AsRef<str>) -> String {
+    let key = key.as_ref();
+
+    if key.starts_with("sys.") {
+        return get_sys_config_str(key);
+    }
+
+    match Config::from_str(key) {
+        Ok(config_key) => {
+            let value = match config_key {
+                Config::Selfavatar => {
+                    let rel_path = context.sql.get_config(context, key, None);
+                    rel_path.map(|p| {
+                        let v = unsafe { dc_get_abs_path(context, to_cstring(p).as_ptr()) };
+                        let r = to_string(v);
+                        unsafe { free(v as *mut _) };
+                        r
+                    })
+                }
+                _ => context.sql.get_config(context, key, None),
+            };
+
+            if value.is_some() {
+                return value.unwrap();
+            }
+
+            // Default values
+            match config_key {
+                Config::Selfstatus => {
+                    let s = unsafe { dc_stock_str(context, 13) };
+                    let res = to_string(s);
+                    unsafe { free(s as *mut _) };
+                    res
+                }
+                _ => config_key
+                    .get_str("default")
+                    .unwrap_or_default()
+                    .to_string(),
+            }
+        }
+        Err(_) => "".into(),
+    }
+}
+
+fn get_sys_config_str(key: impl AsRef<str>) -> String {
+    match SysConfig::from_str(key.as_ref()) {
+        Ok(SysConfig::Version) => std::str::from_utf8(DC_VERSION_STR).unwrap().into(),
+        Ok(SysConfig::MsgsizeMaxRecommended) => format!("{}", 24 * 1024 * 1024 / 4 * 3),
+        Ok(SysConfig::ConfigKeys) => get_config_keys_str(),
+        Err(_) => "".into(),
+    }
+}
+
+fn get_config_keys_str() -> String {
+    let keys = Config::iter().fold(String::new(), |mut acc, key| {
+        acc += key.as_ref();
+        acc += " ";
+        acc
+    });
+
+    let sys_keys = SysConfig::iter().fold(String::new(), |mut acc, key| {
+        acc += key.as_ref();
+        acc += " ";
+        acc
+    });
+
+    format!(" {} {} ", keys, sys_keys)
+}
+
+/// Set the given config key.
+/// Returns `1` on success and `0` on failure.
+pub fn set(context: &Context, key: impl AsRef<str>, value: Option<&str>) -> libc::c_int {
```

Comment:
Can we not make this take the enum?  Doing the from string stuff should only be needed for the C-API compatibility.

Reply:
that would be nicer, I could push this string dancing out of this method